From 0d45eda7e83b9d8f07e709f3efbe7728d025bd26 Mon Sep 17 00:00:00 2001
From: Benoit Parrot <bparrot@ti.com>
Date: Fri, 7 Nov 2014 12:52:24 -0600
Subject: [PATCH 1321/1587] media: ti-vpfe: Fix format enumeration to match
 sensor capabilities

Format enumeration was based on a static/private format table.
Instead we always query the sub device to get valid/active formats.

Signed-off-by: Benoit Parrot <bparrot@ti.com>
[zou: Original patch taken from
ti-sdk-am335x-evm-08.00.00.00-Linux-x86-Install.bin]
Signed-off-by: Cao Zou <cao.zou@windriver.com>
---
 drivers/media/platform/ti-vpfe/vpfe.c |   30 ++++++++++++++++++++++++------
 1 files changed, 24 insertions(+), 6 deletions(-)

diff --git a/drivers/media/platform/ti-vpfe/vpfe.c b/drivers/media/platform/ti-vpfe/vpfe.c
index 108fb46..453cfd0 100644
--- a/drivers/media/platform/ti-vpfe/vpfe.c
+++ b/drivers/media/platform/ti-vpfe/vpfe.c
@@ -108,7 +108,7 @@ static struct vpfe_fmt formats[] = {
 	{
 		.name		= "4:2:2, packed, UYVY",
 		.fourcc		= V4L2_PIX_FMT_UYVY,
-		.code		= V4L2_MBUS_FMT_YUYV8_2X8,
+		.code		= V4L2_MBUS_FMT_UYVY8_2X8,
 		.l.width	= 10,
 		.l.bpp		= 4,
 		.s.width	= 8,
@@ -191,7 +191,7 @@ static struct vpfe_fmt formats[] = {
 	{
 		.name		= "RGB565 (LE)",
 		.fourcc		= V4L2_PIX_FMT_RGB565,
-		.code		= V4L2_MBUS_FMT_BGR565_2X8_LE,
+		.code		= V4L2_MBUS_FMT_RGB565_2X8_LE,
 		.l.width	= 10,
 		.l.bpp		= 4,
 		.s.width	= 8,
@@ -200,7 +200,7 @@ static struct vpfe_fmt formats[] = {
 	{
 		.name		= "RGB565 (BE)",
 		.fourcc		= V4L2_PIX_FMT_RGB565X,
-		.code		= V4L2_MBUS_FMT_BGR565_2X8_BE,
+		.code		= V4L2_MBUS_FMT_RGB565_2X8_BE,
 		.l.width	= 10,
 		.l.bpp		= 4,
 		.s.width	= 8,
@@ -1905,18 +1905,36 @@ static int vpfe_g_fmt(struct file *file, void *priv,
 }
 
 static int vpfe_enum_fmt(struct file *file, void  *priv,
-				   struct v4l2_fmtdesc *f)
+			 struct v4l2_fmtdesc *f)
 {
 	struct vpfe_device *dev = video_drvdata(file);
+	struct v4l2_subdev_mbus_code_enum mbus_code;
+	struct v4l2_subdev *subdev;
+	struct media_pad *remote;
 	struct vpfe_fmt *fmt;
+	u32 pad;
+	int ret;
 
 	vpfe_dbg(2, dev, "vpfe_enum_format index:%d\n",
 		f->index);
 
-	if (f->index >= ARRAY_SIZE(formats))
+	subdev = vpfe_remote_subdev(dev, &pad);
+	if (subdev == NULL)
 		return -EINVAL;
 
-	fmt = &formats[f->index];
+	remote = media_entity_remote_pad(&dev->pad);
+	mbus_code.index = f->index;
+	ret = v4l2_subdev_call(subdev, pad, enum_mbus_code, NULL, &mbus_code);
+	if (ret)
+		return -EINVAL;
+
+	/* convert mbus_format to v4l2_format */
+	fmt = find_format_by_code(mbus_code.code);
+	if (fmt == NULL) {
+		vpfe_err(dev, "mbus code 0x%08x not found\n",
+			mbus_code.code);
+		return -EINVAL;
+	}
 
 	strncpy(f->description, fmt->name, sizeof(f->description) - 1);
 	f->pixelformat = fmt->fourcc;
-- 
1.7.5.4

